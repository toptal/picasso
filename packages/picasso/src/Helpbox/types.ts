export interface HelpboxContextProps {
  closeable?: boolean
}
