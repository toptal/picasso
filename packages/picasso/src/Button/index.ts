import type { Props } from './Button'
export { default } from './Button'
export type ButtonProps = Props
export * from './Button'
