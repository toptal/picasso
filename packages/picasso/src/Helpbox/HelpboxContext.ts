import React from 'react'

import type { HelpboxContextProps } from './types'

const HelpboxContext = React.createContext<HelpboxContextProps>({})

export default HelpboxContext
