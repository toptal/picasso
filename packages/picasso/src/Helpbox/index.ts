import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Helpbox'

export { default } from './Helpbox'
export type HelpboxProps = OmitInternalProps<Props>
export * from './HelpboxContext'
