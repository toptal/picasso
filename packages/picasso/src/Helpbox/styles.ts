import { createStyles } from '@material-ui/core/styles'

export default () =>
  createStyles({
    root: {
      position: 'relative'
    },
    closeButton: {
      position: 'absolute',
      top: '0.875em',
      right: '0.875em',
      background: 'transparent',
      border: 0,
      padding: 0,

      '&:hover': {
        background: 'transparent'
      }
    }
  })
