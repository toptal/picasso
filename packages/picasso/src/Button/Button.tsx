import React, {
  ReactNode,
  ReactElement,
  MouseEvent,
  forwardRef,
  ElementType
} from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import ButtonBase from '@material-ui/core/ButtonBase'
import {
  BaseProps,
  SizeType,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  useAppConfig
} from '@toptal/picasso-shared'

import Loader from '../Loader'
import Container from '../Container'
import Group from '../ButtonGroup'
import kebabToCamelCase from '../utils/kebab-to-camel-case'
import toTitleCase from '../utils/to-title-case'
import styles from './styles'

export type VariantType =
  | 'primary-blue'
  | 'secondary-blue'
  | 'primary-red'
  | 'secondary-red'
  | 'primary-green'
  | 'secondary-green'
  | 'flat'
  | 'flat-white'
  | 'secondary-white'
  | 'transparent'
  | 'transparent-white'
  | 'transparent-blue'

export type IconPositionType = 'left' | 'right'

export interface Props extends BaseProps, ButtonOrAnchorProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** The component used for the root node. Either a string to use a DOM element or a component. */
  as?: ElementType
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
  // TODO: should it be exposed?
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  /** Set hovered style for the button */
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** Shows a loading indicator and disables click events */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => void
  /** A button can have different sizes */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** The variant to use */
  variant?: VariantType
  /** HTML Value of Button component */
  value?: string | number
  /** Circular style of Button component */
  circular?: boolean
  /** HTML title of Button component */
  title?: string
  /** HTML type of Button component */
  type?: 'button' | 'reset' | 'submit'
}

interface StaticProps {
  Group: typeof Group
}

const getVariantType = (variant: VariantType) => {
  const [type] = variant!.split('-')

  return type
}

const useStyles = makeStyles<Theme, Props>(styles, { name: 'PicassoButton' })

const defaultOnClick = () => {}

export const Button = forwardRef<HTMLButtonElement, Props>(function Button(
  props,
  ref
) {
  const {
    icon,
    iconPosition,
    loading,
    children,
    className,
    style,
    fullWidth,
    variant,
    size,
    focused,
    hovered,
    disabled,
    active,
    onClick,
    circular,
    title,
    value,
    type,
    as,
    ...rest
  } = props
  const classes = useStyles(props)

  const { titleCase } = useAppConfig()

  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
    root: rootClass,
    hidden: hiddenClass,
    loader: loaderClass,
    content: contentClass
  } = classes

  const finalChildren = [titleCase ? toTitleCase(children) : children]

  if (icon) {
    const iconComponent = React.cloneElement(icon, {
      className: cx(iconClass, icon.props.className, {
        [iconLeftClass]: children && iconPosition === 'left',
        [iconRightClass]: children && iconPosition === 'right'
      }),
      key: 'button-icon'
    })

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantType = getVariantType(variant!)
  const variantClassName = disabled
    ? classes[`${variantType}Disabled`]
    : classes[kebabToCamelCase(variant!)]
  const sizeClassName = classes[size!]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered,
      [classes.circular]: circular
    },
    sizeClassName,
    variantClassName,
    rootClass
  )

  return (
    <ButtonBase
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={ref}
      classes={{
        root: rootClassName
      }}
      onClick={loading ? defaultOnClick : onClick}
      className={className}
      style={style}
      disabled={disabled}
      title={title}
      value={value}
      type={type}
      component={as!}
    >
      <Container
        as='span'
        inline
        flex
        direction='row'
        alignItems='center'
        className={cx({ [hiddenClass]: loading }, contentClass)}
      >
        {finalChildren}
      </Container>

      {loading && (
        <Loader variant='inherit' className={loaderClass} inline size='small' />
      )}
    </ButtonBase>
  )
}) as CompoundedComponentWithRef<Props, HTMLButtonElement, StaticProps>

Button.defaultProps = {
  active: false,
  as: 'button',
  children: null,
  circular: false,
  disabled: false,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: defaultOnClick,
  size: 'medium',
  type: 'button',
  variant: 'primary-blue'
}

Button.displayName = 'Button'

Button.Group = Group

export default Button as OverridableComponent<Props> & StaticProps
