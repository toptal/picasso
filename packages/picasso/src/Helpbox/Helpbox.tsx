import React, { ReactNode, forwardRef, HTMLAttributes } from 'react'
import { makeStyles, Theme } from '@material-ui/core/styles'
import cx from 'classnames'
import { BaseProps } from '@toptal/picasso-shared'

import Container, { VariantType as ContainerVariantType } from '../Container'
import HelpboxTitle from '../HelpboxTitle'
import HelpboxContent from '../HelpboxContent'
import HelpboxActions from '../HelpboxActions'
import { CloseMinor16 } from '../Icon'
import ButtonCircular from '../ButtonCircular'
import styles from './styles'
import HelpboxContext from './HelpboxContext'

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** Children components (`Helpbox.Title`, `Helpbox.Content`, `Hdlpbox.Actions`) */
  children: ReactNode
  /** Color variant of Helpbox */
  variant?: ContainerVariantType
  /** Callback invoked when close is clicked */
  onClose?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void
}

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoHelpbox'
})

// eslint-disable-next-line react/display-name
export const Helpbox = forwardRef<HTMLDivElement, Props>(function Helpbox(
  props,
  ref
) {
  const { className, style, children, variant, onClose, ...rest } = props
  const classes = useStyles()

  return (
    <Container
      {...rest}
      ref={ref}
      rounded
      className={cx(classes.root, className)}
      style={style}
      bordered={!variant || variant === 'white'}
      variant={variant}
      padded='medium'
    >
      <HelpboxContext.Provider value={{ closeable: Boolean(onClose) }}>
        {children}
      </HelpboxContext.Provider>
      {onClose && (
        <ButtonCircular
          className={classes.closeButton}
          onClick={onClose}
          icon={<CloseMinor16 color='dark-grey' />}
        />
      )}
    </Container>
  )
})

Helpbox.displayName = 'Helpbox'

export default Object.assign(Helpbox, {
  Title: HelpboxTitle,
  Content: HelpboxContent,
  Actions: HelpboxActions
})
