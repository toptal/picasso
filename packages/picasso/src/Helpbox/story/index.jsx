import helpboxTitleStory from '../../HelpboxTitle/story'
import helpboxContentStory from '../../HelpboxContent/story'
import helpboxActionsStory from '../../HelpboxActions/story'
import { Helpbox } from '../Helpbox'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Helpbox',
  'Container specialized for rendering suggestions'
)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: Helpbox, name: 'Helpbox' })
  .addComponentDocs(helpboxTitleStory.componentDocs)
  .addComponentDocs(helpboxContentStory.componentDocs)
  .addComponentDocs(helpboxActionsStory.componentDocs)

page
  .createChapter()
  .addExample('Helpbox/story/Default.example.tsx', 'Default')
  .addExample('Helpbox/story/Actions.example.tsx', 'With actions')
  .addExample('Helpbox/story/Closeable.example.tsx', 'Closeable')
