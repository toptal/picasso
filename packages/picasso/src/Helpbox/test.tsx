import React from 'react'
import { render } from '@toptal/picasso/test-utils'
import { OmitInternalProps } from '@toptal/picasso-shared'

import Helpbox, { Props } from './Helpbox'
import Button from '../Button'

const TestHelpbox = ({ children }: OmitInternalProps<Props>) => {
  return (
    <Helpbox>
      <Helpbox.Title>Title</Helpbox.Title>
      <Helpbox.Content>{children}</Helpbox.Content>
      <Helpbox.Actions>
        <Button>Button</Button>
      </Helpbox.Actions>
    </Helpbox>
  )
}

describe('Helpbox', () => {
  it('renders', () => {
    const { container } = render(<TestHelpbox>Test</TestHelpbox>)

    expect(container).toMatchSnapshot()
  })
})
