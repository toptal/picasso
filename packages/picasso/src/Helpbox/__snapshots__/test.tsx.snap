// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpbox renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoContainer-mediumPadding PicassoContainer-bordered PicassoContainer-rounded PicassoHelpbox-root"
    >
      <div
        class="PicassoContainer-bottomsmallMargin"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
        >
          Title
        </h4>
      </div>
      <div
        class="MuiTypography-root PicassoTypography-bodyMedium PicassoTypography-black MuiTypography-body1"
      >
        Test
      </div>
      <div
        class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoHelpboxActions-root"
      >
        <button
          class="MuiButtonBase-root PicassoButton-medium PicassoButton-primary PicassoButton-root"
          data-component-type="button"
          tabindex="0"
          type="button"
        >
          <span
            class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
          >
            Button
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
